src/FSharp.Core/printf.fs

```diff
@@ -658,8 +658,28 @@ module internal PrintfImpl =
     
     /// Contains functions to handle left/right and no justification case for numbers
     module GenericNumber =
-
-        let isPositive (n: obj) = 
+        
+        let inline doubleIsPositive (n: double) =
+            n >= 0.0
+                // Ensure -0.0 is treated as negative (see https://github.com/dotnet/fsharp/issues/15557)
+                // and use bitwise comparison because floating point comparison treats +0.0 as equal to -0.0
+                && (BitConverter.DoubleToInt64Bits n <> BitConverter.DoubleToInt64Bits -0.0)
+        
+        let inline singleIsPositive (n: single) =
+            doubleIsPositive (float n)
+        
+        let decimalSignBit (n: decimal) =
+            // Unfortunately it's impossible to avoid this array allocation without either targeting .NET 5+ or relying on knowledge about decimal's internal representation
+            let bits = Decimal.GetBits n
+            bits[3] >>> 31
+        
+        let inline decimalIsNegativeZero (n: decimal) =
+            decimalSignBit n <> 0
```

Comment:
This is notably checking `IsNegative`, not `IsNegativeZero` and would also return true for `-1.0m`, as an example.